File: session-2/nodejs/consumer_salon.js
#!/usr/bin/env node
/**
 * Consumer Salon - Recoit uniquement les messages du salon
 *
 * Binding key : maison.salon.*
 * Le wildcard * matche exactement un mot, donc ce consumer recoit tous les
 * capteurs du salon (temperature, mouvement, fumee, etc.)
 */

const amqp = require('amqplib');

const EXCHANGE_NAME = 'domotique';
const BINDING_KEY = 'maison.salon.*';

async function main() {
  // Connexion a RabbitMQ
  const connection = await amqp.connect('amqp://localhost');
  const channel = await connection.createChannel();

  // Declaration du Topic Exchange "domotique"
  await channel.assertExchange(EXCHANGE_NAME, 'topic', { durable: false });

  // Creation d'une queue exclusive (nom genere automatiquement, auto-delete)
  const q = await channel.assertQueue('', { exclusive: true });

  // Binding : on s'abonne uniquement aux messages du salon
  await channel.bindQueue(q.queue, EXCHANGE_NAME, BINDING_KEY);

  console.log(`[*] Consumer SALON demarre (binding: ${BINDING_KEY})`);
  console.log('[*] En attente des messages du salon... Ctrl+C pour arreter.\n');

  // Consommation des messages
  channel.consume(q.queue, (msg) => {
    if (msg) {
      const message = JSON.parse(msg.content.toString());
      const sensor = message.sensor || 'inconnu';
      const value = message.value !== undefined ? message.value : 'N/A';

      // Affichage adapte au type de capteur
      if (sensor === 'thermometre') {
        console.log(`[SALON] Temperature: ${value}°C`);
      } else if (sensor === 'detecteur_mouvement') {
        console.log(`[SALON] Mouvement detecte: ${value}`);
      } else if (sensor === 'detecteur_fumee') {
        console.log(`[SALON] Fumee: niveau ${value}`);
      } else {
        console.log(`[SALON] ${sensor}: ${value}`);
      }
    }
  }, { noAck: true });

  // Gestion de l'arret propre
  process.on('SIGINT', () => {
    console.log('\n[*] Arret du consumer salon.');
    connection.close();
    process.exit(0);
  });
}

main().catch(console.error);

File: session-2/python/consumer_salon.py
#!/usr/bin/env python3
"""
Consumer Salon - Recoit uniquement les messages du salon

Binding key : maison.salon.*
Le wildcard * matche exactement un mot, donc ce consumer recoit tous les
capteurs du salon (temperature, mouvement, fumee, etc.)
"""

import pika
import json

# Connexion a RabbitMQ
connection = pika.BlockingConnection(
    pika.ConnectionParameters('localhost')
)
channel = connection.channel()

# Declaration du Topic Exchange "domotique"
channel.exchange_declare(
    exchange='domotique',
    exchange_type='topic'
)

# Creation d'une queue exclusive (nom genere automatiquement, auto-delete)
result = channel.queue_declare(queue='', exclusive=True)
queue_name = result.method.queue

# Binding : on s'abonne uniquement aux messages du salon
binding_key = 'maison.salon.*'
channel.queue_bind(
    exchange='domotique',
    queue=queue_name,
    routing_key=binding_key
)

print(f"[*] Consumer SALON demarre (binding: {binding_key})")
print("[*] En attente des messages du salon... Ctrl+C pour arreter.\n")


def callback(ch, method, properties, body):
    """Callback appele a chaque message recu"""
    message = json.loads(body)
    sensor = message.get('sensor', 'inconnu')
    value = message.get('value', 'N/A')

    # Affichage adapte au type de capteur
    if sensor == 'thermometre':
        print(f"[SALON] Temperature: {value}°C")
    elif sensor == 'detecteur_mouvement':
        print(f"[SALON] Mouvement detecte: {value}")
    elif sensor == 'detecteur_fumee':
        print(f"[SALON] Fumee: niveau {value}")
    else:
        print(f"[SALON] {sensor}: {value}")


# Consommation des messages (auto_ack=True pour simplifier)
channel.basic_consume(
    queue=queue_name,
    on_message_callback=callback,
    auto_ack=True
)

try:
    channel.start_consuming()
except KeyboardInterrupt:
    print("\n[*] Arret du consumer salon.")
    connection.close()
